### ISSUE TITLE
Multiple oauth2-proxy instances in k8s - failing due to missing CSRF token

### ISSUE THREAD

**Issue by charvadzo** [oauth2-proxy/oauth2-proxy#2256]
## Expected Behavior

It should be possible to reliably perform oauth2 authentication flow even when running multiple oauth2-proxy instances in k8s with random loadbalancing of requests between them.

## Current Behavior

When running a single oauth2-proxy instance the authentication works reliably. But when running 2 or more instances the authentication often fails, throwing 403 on the callback request with message "Login Failed: Unable to find a valid CSRF token. Please try again." 

We're using Istio for managing ingress to our cluster and oauth2-proxy is configured as it's external authorization provider.

From the logs of oauth2-proxy instances there is apparent that the authentication flow alternates between the two instances which is likely causing the issue:

```
oauth2-proxy instance A:

xxx-oauth2-proxy [2023/10/02 11:13:11] [oauthproxy.go:959] No valid authentication in request. Initiating login.
xxx-oauth2-proxy 127.0.0.6 - 888976d7-2b40-4777-8688-b4b4b8cef7e2 - - [2023/10/02 11:13:11] our.domain GET - "/favicon.ico" HTTP/1.1 "" 302 419 0.000
xxx-oauth2-proxy 127.0.0.6 - 534ef1d3-0469-4e98-aaeb-7452f05cb347 - my@email [2023/10/02 11:13:11] [AuthSuccess] Authenticated via OAuth2: Session{email:my@email user: PreferredUsername:my.user token:true id_token:true created:2023-10-02 11:13:11.32483328 +0000 UTC m=+367966.812581108 expires:2023-10-02 11:18:11.324601649 +0000 UTC m=+368266.812349474 refresh_token:true groups:[role:default-roles-xxx role:offline_access role:uma_authorization role:xxx:xxx-web-gui role:account:manage-account role:account:manage-account-links role:account:view-profile]}
xxx-oauth2-proxy [2023/10/02 11:13:11] [session_store.go:163] WARNING: Multiple cookies are required for this session as it exceeds the 4kb cookie limit. Please use server side session storage (eg. Redis) instead.
xxx-oauth2-proxy 127.0.0.6 - 534ef1d3-0469-4e98-aaeb-7452f05cb347 - - [2023/10/02 11:13:11] our.domain GET - "/oauth2/callback?state=pYNre0PYkglD7o-ZoermlyPhFsk9WKwI6VfOdtA4S1w%3A%2Ffavicon.ico&session_state=d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4&code=8d1b47e0-abc2-41c6-8533-6b936bd452ef.d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4.b7d27d00-1cf6-4e30-81e2-409c27f9824d" HTTP/1.1 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36" 302 35 0.039

oauth2-proxy instance B:

xxx-oauth2-proxy [2023/10/02 11:13:10] [oauthproxy.go:816] &{GET /oauth2/callback?state=a9RAZ175hPUh6Nxrf94ec83zMqs-10R4L8Xdi9GemR8%3A%2F&session_state=d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4&code=ad13d3cb-9b17-41b6-b5aa-00084eeda2fe.d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4.b7d27d00-1cf6-4e30-81e2-409c27f9824d HTTP/1.1 1 1 map[Accept:[text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7] Accept-Encoding:[gzip, deflate, br] Accept-Language:[en-US,en;q=0.9,cs-CZ;q=0.8,cs;q=0.7] Cache-Control:[max-age=0] Sec-Ch-Ua:["Google Chrome";v="117", "Not;A=Brand";v="8", "Chromium";v="117"] Sec-Ch-Ua-Mobile:[?0] Sec-Ch-Ua-Platform:["Windows"] Sec-Fetch-Dest:[document] Sec-Fetch-Mode:[navigate] Sec-Fetch-Site:[same-site] Sec-Fetch-User:[?1] Upgrade-Insecure-Requests:[1] User-Agent:[Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36] X-Amzn-Trace-Id:[Root=1-651aa5c6-799501c5799632ab1e1877e2] X-B3-Parentspanid:[45c2155b985774a4] X-B3-Sampled:[0] X-B3-Spanid:[1daf7c78d1ff3fd1] X-B3-Traceid:[87a9bc2451894ac545c2155b985774a4] X-Envoy-Attempt-Count:[1] X-Envoy-External-Address:[46.252.225.100] X-Forwarded-Client-Cert:[By=spiffe://cluster.local/ns/dmp/sa/xxx-oauth2-proxy;Hash=229f0fde0e889104ce56a2a854add0e17f0e5728951730e756a922a249a98b59;Subject="";URI=spiffe://cluster.local/ns/istio-system/sa/ingressgateway] X-Forwarded-For:[46.252.225.100,10.205.12.137] X-Forwarded-Port:[443] X-Forwarded-Proto:[https] X-Request-Id:[81d9b991-1639-40e1-b123-faf550fb86af]] {} <nil> 0 [] false our.domain map[code:[ad13d3cb-9b17-41b6-b5aa-00084eeda2fe.d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4.b7d27d00-1cf6-4e30-81e2-409c27f9824d] session_state:[d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4] state:[a9RAZ175hPUh6Nxrf94ec83zMqs-10R4L8Xdi9GemR8:/]] map[] <nil> map[] 127.0.0.6:49533 /oauth2/callback?state=a9RAZ175hPUh6Nxrf94ec83zMqs-10R4L8Xdi9GemR8%3A%2F&session_state=d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4&code=ad13d3cb-9b17-41b6-b5aa-00084eeda2fe.d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4.b7d27d00-1cf6-4e30-81e2-409c27f9824d <nil> <nil> <nil> 0xc0004a4360} AuthFailure Invalid authentication via OAuth2: unable to obtain CSRF cookie
xxx-oauth2-proxy 
xxx-oauth2-proxy 127.0.0.6 - 81d9b991-1639-40e1-b123-faf550fb86af - - [2023/10/02 11:13:10] our.domain GET - "/oauth2/callback?state=a9RAZ175hPUh6Nxrf94ec83zMqs-10R4L8Xdi9GemR8%3A%2F&session_state=d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4&code=ad13d3cb-9b17-41b6-b5aa-00084eeda2fe.d0ee53f0-16d0-41e8-ba70-4a7a7e9662e4.b7d27d00-1cf6-4e30-81e2-409c27f9824d" HTTP/1.1 "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36" 403 2800 0.000
xxx-oauth2-proxy 127.0.0.6 - 1d92402c-2c8b-4757-8fb0-5b9a197ab2eb - - [2023/10/02 11:13:11] 10.205.11.57:4180 GET - "/ping" HTTP/1.1 "kube-probe/1.24+" 200 2 0.000
xxx-oauth2-proxy 127.0.0.6 - 94b4a951-9407-4702-9a81-2b87408aa222 - my@email [2023/10/02 11:13:11] our.domain GET static://200 "/favicon.ico" HTTP/1.1 "" 200 13 0.000
```

## Steps to Reproduce (for bugs)

1.  deploy oauth2-proxy to k8s as a deployment with 2 or more replicas
2.  setup Istio to use oauth2-proxy as an extension provider, setup virtual service,...
3.  try to perform oauth2 flow several times (the more replicas you set the better chance the problem occurs)

## Context

We hope that it should be possible to scale out oauth2-proxy without using any sticky sessions etc. as it is stateless (unless using Redis for token storage). Maybe we have something set up wrong. Tried --cookie-csrf-per-request=true with no luck.

## Your Environment

- AWS EKS cluster
- Istio for ingress control and mTLS with oauth2-proxy as extension provider
- oauth2-proxy deployed using bitnami helm chart with 2 replica set

## Version used:
7.4.0-debian-11-r57

**Comment by JoelSpeed** [comment]
I've run multiple OAuth2 Proxy instances and had requests flit between them before without issue. I suspect this is a configuration issue since there is no state stored for a particular request within memory.
Normally the cookies are the only state, and those don't matter re-which backend they hit.

Do you have identical configuration for both proxies? In particular, is the cookie secret the same between the two?

**Comment by charvadzo** [comment]
Hi JoelSpeed,
I'm glad to hear you were able to run multiple instances without issues. Then It can be that something is wrong on our end.

Both proxies are deployed from k8s deployment manifest so receiving identical configuration:

```
--provider=keycloak-oidc 
--oidc-issuer-url=<...>
--http-address=0.0.0.0:4180
--upstream=static://200
--reverse-proxy=true
--force-https=false
--set-authorization-header=true
--cookie-secure=true
--cookie-samesite=lax
--cookie-refresh=3m
--cookie-expire=1h
--cookie-name=<...>
--email-domain=*
--skip-provider-button=true
--whitelist-domain=<...>
--skip-jwt-bearer-tokens=true
--cookie-domain=.<...>
```
The cookie secret is set to env variable from k8s secret which is common to both instances as well.
And if the secret was wrong then there should be "CSRF cookie failed validation" message in the oauth2-proxy logs which is not true.

We're thinking of tryout with Redis option as a session storage instead but would be happy if we find out where is the issue with the cookies.

**Comment by charvadzo** [comment]
Hi JoelSpeed,

so you were right :-)
Thanks for pointing us in the right direction.

### TL;DR

The issue was that in some cases the oauth2-proxy instances did not have the same cookie secret configured.


### Additional info

We're deploying the oauth2-proxy using our custom Helm chart where the cookie is generated randomly by Helm using randAlphaNum function. The issue is that this function regenerates the secret at every Helm upgrade, even if the upgrade does not change anything in the oauthh2-proxy (sub)chart.

That said, any upgrade was causing the secret regeneration but the oauth2-proxy instances might not have restarted during the upgrade thus remained configured with the old secret. Then if one of the instances have been restarted for some reason, it loaded the new secret while rest of the instances kept the old one. With this setup, the CSRF cookie could be encoded by an instance using the old secret but decoded from a subsequent request by another instance using the new secret, thus failing validation.

An issue during debugging was that on a failed validation the oauth2-proxy should generate an error "CSRF cookie failed validation" which was true but the error was not logged with this message so I've mistakenly factored out this possible cause. The only option how to see this error message is to turn on showing debug information on error page (--show-debug-on-error). This option, when discovered, was later used to verify that this is the cause of the problem.


### Question

Would be a small PR with logging any error returned by cookies.LoadCSRFCookie method welcomed? That would have saved me several days of debugging. I understand it should not be showed on the error page by default but in the logs I assume it should not be of a security concern, or?
